src/node/dynamics_compressor.rs

```diff
@@ -0,0 +1,512 @@
+        if self.ring_buffer.len() < ring_size {
+            let mut silence = input.clone();
+            silence.make_silent();
+            self.ring_buffer.resize(ring_size, silence);
+        }
+
+        // setup values for compression curve
+        // https://webaudio.github.io/web-audio-api/#compression-curve
+        let threshold = params.get(&self.threshold)[0];
+        let knee = params.get(&self.knee)[0];
+        let ratio = params.get(&self.ratio)[0];
+        // @note: if knee != 0. we shadow threshold to match definitions of knee
+        //   and threshold given in https://www.eecs.qmul.ac.uk/~josh/documents/2012/
+        //   where knee is centered around threshold.
+        //   We can thus reuse their formula for the gain computer stage.
+        // yG =
+        //     xG                                      if 2(xG − T) < −W
+        //     xG + (1/R − 1)(xG − T + W/2)^2 / (2W)   if 2|(xG − T)| ≤ W
+        //     T + (xG − T)/R                          if 2(xG − T) > W
+        // This is weird, and probably wrong because `knee` and `threshold` are not
+        // independant, but matches the spec.
+        let threshold = if knee > 0. {
+            threshold + knee / 2.
+        } else {
+            threshold
+        };
+        let half_knee = knee / 2.;
+        // pre-compute for this block the constant part of the formula of the knee
+        let knee_partial = (1. / ratio - 1.) / (2. * knee);
+
+        // compute time constants for attack and release - eq. (7) in paper
+        let attack = params.get(&self.attack)[0];
+        let release = params.get(&self.release)[0];
+        let attack_tau = (-1. / (attack * sample_rate)).exp();
+        let release_tau = (-1. / (release * sample_rate)).exp();
+
+        // Computing the makeup gain means executing the following steps:
+        // - Let full range gain be the value returned by applying the compression curve to the value 1.0.
+        // - Let full range makeup gain be the inverse of full range gain.
+        // - Return the result of taking the 0.6 power of full range makeup gain.
+        // @note: this should be confirmed / simplified, maybe could do all this in dB
+        // seems coherent with chrome implementation
+        let full_range_gain = threshold + (-threshold / ratio);
+        let full_range_makeup = 1. / db_to_lin(full_range_gain);
+        let makeup_gain = lin_to_db(full_range_makeup.powf(0.6));
+
+        let mut prev_detector_value = self.prev_detector_value;
+
+        let mut reduction_gain = 0.; // dB
+        let mut reduction_gains = [0.; 128]; // lin
+        let mut detector_values = [0.; 128]; // lin
+
+        for i in 0..RENDER_QUANTUM_SIZE {
+            // pick highest value for this index across all input channels
+            // @tbc - this seems to be what is done in chrome
+            let mut max = f32::MIN;
+
+            for channel in input.channels().iter() {
+                let sample = channel.as_slice()[i];
```

Comment:
Should you not take the `.abs()` already here? Otherwise you would only consider positive values

Reply:
Sure, good catch!